### PR TITLE
update for run 2017

### PR DESCRIPTION
**Pull request by Somhammer**
1. plugin/ttbbLepJetAnalyzer.cc
  - CSV -> deepCSV
  - Electron ID (80X -> 94X)
  - Prefire weight

2. python/lepton_SF.py
  - Add electron SF ( ID + Reco + Zvtx )

### CODE REVIEW

File: CatAnalyzer/prod/ttbbLepJetsAnalyzer_cfg.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    print "Running data.... Including JSON File."
+    import FWCore.PythonUtilities.LumiList as LumiList
+    process.source.lumisToProcess = LumiList.LumiList(filename = '../../CatProducer/data/LumiMask/Cert_294927-306462_13TeV_EOY2017ReReco_Collisions17_JSON.txt').getVLuminosityBlockRange()
+
+# Lepton Scale Factors
+from CATTools.CatAnalyzer.leptonSF_cff import *
+# GEN Weights
+process.load("CATTools.CatAnalyzer.flatGenWeights_cfi")
+# CSV Scale Factors
+process.load("CATTools.CatAnalyzer.csvWeights_cfi")
```
Comment (minerva1993) [inline CatAnalyzer/prod/ttbbLepJetsAnalyzer_cfg.py]: This SF is based on ttH SF root file, which is not b POG recipe. I strongly recommend you to remove this part. Also, the config for deepCSV is following one: https://github.com/vallot/CATTools/blob/cat90x/CatAnalyzer/python/deepcsvWeights_cfi.py
csvWeights_cfi is for csv v2.

Also it is recommended to remove deepcsvWeightLabel.

File: CatAnalyzer/python/leptonSF_cff.py [hunk 2]
```diff
@@ -834,5 +834,6 @@ def combineSF(set1, set2, additionalUnc1=0, additionalUnc2=0):
 #2017 SF - ongoing
 muonSFTight94X = combineSF(muonSFTightIdOnly94X, muonSFTightIsoOnly94X)
 electronSFMVAWP8094Xv2 = combineSF( combineSF( electronSFMVAWP80IsoIdOnly94Xv2, electronSFMVAWP80RecoOnly94Xv2 ), electronSFHLTZvtx94X)
+electronSFCutBasedIDTightWP94X = combineSF( combineSF( electronSFCutBasedTightIDOnly94Xv2, electronSFMVAWP80RecoOnly94Xv2), electronSFHLTZvtx94X)
```
Comment (minerva1993) [inline CatAnalyzer/python/leptonSF_cff.py]: electronSFCutBasedIDTightWP94X includes Reco, Zvtx SF which is misleading. Change name or remove it.

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 3]
```diff
@@ -85,11 +77,13 @@ class ttbbLepJetsAnalyzer : public edm::one::EDAnalyzer<edm::one::SharedResource
...
   edm::EDGetTokenT<float>                        puWeightToken_;
   edm::EDGetTokenT<float>                        puUpWeightToken_;
   edm::EDGetTokenT<float>                        puDownWeightToken_;
-  edm::EDGetTokenT<float>                        CSVWeightToken_;
-  edm::EDGetTokenT<std::vector<float>>           CSVSysWeightToken_;
+  edm::EDGetTokenT<double>                       prefWeightToken_;
+  edm::EDGetTokenT<double>                       prefWeightUpToken_;
+  edm::EDGetTokenT<double>                       prefWeightDownToken_;
   // Object Collections
   edm::EDGetTokenT<reco::GenParticleCollection>  genToken_;
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: GenJets and GenParticle collections are no longer stored in cattuple.

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 4]
```diff
@@ -282,33 +246,47 @@ ttbbLepJetsAnalyzer::ttbbLepJetsAnalyzer(const edm::ParameterSet& iConfig):
...
+  puDownWeightToken_    = consumes<float>             (edm::InputTag(puWeightLabel.label(),"dn"));
+  // Prefie (16 and 17 only)
+  prefWeightToken_      = consumes<double>(edm::InputTag("prefiringweight:nonPrefiringProb"));
+  prefWeightUpToken_    = consumes<double>(edm::InputTag("prefiringweight:nonPrefiringProbUp"));
+  prefWeightDownToken_  = consumes<double>(edm::InputTag("prefiringweight:nonPrefiringProbDown"));
   // CSV Weights
-  auto csvWeightLabel = iConfig.getParameter<edm::InputTag>("csvWeightLabel");
-  CSVWeightToken_        = consumes<float>             (edm::InputTag(csvWeightLabel.label()));
-  CSVSysWeightToken_     = consumes<std::vector<float>>(edm::InputTag(csvWeightLabel.label(), "syst"));
+  auto deepcsvWeightLabel = iConfig.getParameter<edm::InputTag>("deepcsvWeightLabel");
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc:292]: deprecated recipe.

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 5]
```diff
@@ -394,140 +349,132 @@ ttbbLepJetsAnalyzer::ttbbLepJetsAnalyzer(const edm::ParameterSet& iConfig):
   tree->Branch("run",        &b_Run,         "run/I");
   tree->Branch("luminumber", &b_Lumi_Number, "luminumber/I");
   tree->Branch("genweight",  &b_GenWeight,   "genweight/F");
+  tree->Branch("TruePV",     &b_nTruePV,     "TruePV/I");
   tree->Branch("GoodPV",     &b_nGoodPV,     "GoodPV/I");
   tree->Branch("channel",    &b_Channel,     "channel/I");
+  tree->Branch("eeprefire",  &b_eeprefire, "eeprefire/I");
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: This should be removed, prefire weight is the newest and recommended recipe.

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 6]
```diff
@@ -394,140 +349,132 @@ ttbbLepJetsAnalyzer::ttbbLepJetsAnalyzer(const edm::ParameterSet& iConfig):
...
+  tree->Branch("jet_e" ,            "std::vector<float>", &b_Jet_e);
+  tree->Branch("jet_index" ,        "std::vector<int>",   &b_Jet_Index);
+  tree->Branch("jet_deepCSV" ,      "std::vector<float>", &b_Jet_deepCSV);
+  tree->Branch("jet_deepJet" ,      "std::vector<float>", &b_Jet_deepJet);
+  tree->Branch("jet_SF_deepCSV_30", "std::vector<float>", &b_Jet_SF_deepCSV_30);
+  tree->Branch("jet_SF_deepJet_30", "std::vector<float>", &b_Jet_SF_deepJet_30);
+  tree->Branch("jet_deepCvsL",      "std::vector<float>", &b_Jet_deepCvsL);
+  tree->Branch("jet_deepCvsB",      "std::vector<float>", &b_Jet_deepCvsB);
+  tree->Branch("jet_deepJetCvsL",   "std::vector<float>", &b_Jet_deepJetCvsL);
+  tree->Branch("jet_deepCvsB",      "std::vector<float>", &b_Jet_deepJetCvsB);
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: jet_deepCvsB -> jet_deepJetCvsB

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 7]
```diff
@@ -1208,14 +1063,17 @@ void ttbbLepJetsAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetu
     b_Lepton_relIso = selectedElectrons.at(0).relIso(0.3);
     b_Lepton_isIso = (b_Lepton_relIso < (std::abs(selectedElectrons[0].scEta()) <= 1.479 ? 0.0766 : 0.0678) );
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: https://twiki.cern.ch/twiki/bin/view/CMS/CutBasedElectronIdentificationRun2#Offline_selection_criteria_for_V
Please update

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 8]
```diff
@@ -1310,81 +1168,65 @@ void ttbbLepJetsAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetu
...
-    for (unsigned int icsv = 0; icsv < rsysSF_CSV->size() ; icsv++)  Jet_SF_CSV[4][icsv+1] = rsysSF_CSV->at(icsv); 
-
-    std::vector<cat::Jet> selectedJets;
-
     // Run again over all Jets (CSV order)
     for (unsigned int i = 0; i < JetIndex.size() ; i++) {
-
       const cat::Jet & jet = jets->at(JetIndex[i]);
 
+      // EE L1. Pre-firing bit, pT > 100 and 2.25 < |eta| < 3.0
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: This part is outdated, should be removed

File: CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc [hunk 9]
```diff
@@ -1629,7 +1278,7 @@ bool ttbbLepJetsAnalyzer::IsSelectMuon(const cat::Muon & i_muon_candidate)
   GoodMuon &= (i_muon_candidate.isTightMuon());
```
Comment (minerva1993) [inline CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc]: POG recipe: https://github.com/vallot/CATTools/blob/cat90x/CatAnalyzer/plugins/fcncLepJetsAnalyzer.cc#L1073

**Review by minerva1993 — COMMENTED** [review]
Here's the first set of comments

File: CatAnalyzer/prod/ttbbLepJetsAnalyzer_cfg.py [hunk 10]
```diff
@@ -64,9 +64,7 @@
 # GEN Weights
 process.load("CATTools.CatAnalyzer.flatGenWeights_cfi")
 # CSV Scale Factors
-process.load("CATTools.CatAnalyzer.csvWeights_cfi")
-process.csvWeights.minPt = 30   # Same cuts than jet selection
-process.csvWeights.maxEta = 2.4
+process.load("CATTools.CatAnalyzer.deepcsvWeights_cfi")
```
Comment (minerva1993) [inline CatAnalyzer/prod/ttbbLepJetsAnalyzer_cfg.py]: "I strongly recommend you to remove this part." from the last comment.

**Review by minerva1993 — COMMENTED** [review]
duplicated recipe for SF. This even not exist in cat10x branch. Why do you keep including it? And this SF is old version, bPOG and ttH group provided new numbers.

File: CatAnalyzer/python/leptonSF_cff.py (hunk 2 above)
Comment (Somhammer) [reply CatAnalyzer/python/leptonSF_cff.py]: I change name as electronCombinedSFTightWP94X.

**Comment by minerva1993** [comment]
https://github.com/vallot/CATTools/blob/69b6697c8eeaeda0305141f311546fdb2a7d3b20/CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc#L1318
-> |eta | < 2.4, not 2.1

https://github.com/vallot/CATTools/blob/69b6697c8eeaeda0305141f311546fdb2a7d3b20/CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc#L1270
As well as electron...

https://github.com/vallot/CATTools/blob/69b6697c8eeaeda0305141f311546fdb2a7d3b20/CatAnalyzer/plugins/ttbbLepJetsAnalyzer.cc#L1151
LES is not clear where it comes from. Rather, I included lepton scale&smearing in V9_6 and V10_2 cattuple like this: https://github.com/vallot/CATTools/blob/cat90x/DataFormats/interface/Electron.h#L78-L84
setSmearedScale() should be applied for BOTH data and mc, on the four vector (changing energy)
Systematic uncertainty likely to be small, and it should be applied in the same way as JEC/JER unc. (which are quite annoying)

Also, as I mentioned, some of SFs are likely to be correlated across the years. For safety I recommend you to split lepton SF source by source.

Furthermore, please upload job script (should be something like job_ttbb.sh) so that it can be centrally stored.

**Comment by Somhammer** [comment]
To reduce the time to make input array, I remain |eta| < 2.1.
If many ttbb events are existed in 2.1 < |eta| < 2.4, I will change eta cut.
And I will update soon HLT, lepton SF,  LES parts.

**Comment by minerva1993** [comment]
Lepton eta was restricted to |eta| < 2.1 mainly due to eta restriction of lepton trigger. This is mentioned in the one of KCMS top-higgs working meeting, probably last year. Since the statistics of ttbb sample is a huge problem in 17~18, it would be better to include those region

**Comment by Somhammer** [comment]
Okay, I changed eta cut.

**Comment by minerva1993** [comment]
I think this PR is about to be merged. Do you have anything to revise? @Somhammer

**Pull request merged**